### ISSUE TITLE
Possibly add scratch space for inference method

### ISSUE THREAD

**Issue by tschaffter** [Sage-Bionetworks/DigitalMammographyChallenge#631]
## Problem

Regarding inference submissions, participants mentioned that their container could include several models and that the remaining space may not be enough for normal operation. Participants also suggest to increase the maximum size of the docker containers that is currently limited to 10 GB.

https://www.synapse.org/#!Synapse:syn4224222/discussion/threadId=1295

This program should not occur if the participant mount a trained model hosted on Synapse (limited to 20 GB in Round 3). However, participants are allowed to pack submission containers with trained model already inside it.

## Solution
I'm not for increasing the default max size of the docker containers. First, it's the default size specified by Docker and running larger container requires to increase the settings OK for us but may be a source of concern for other docker setups (weak reason). It's also an issue if participants pack in dozens of model that takes ages to run on the scoring data. max 10 GB / container should be reasonable especially since they can use the trained model hosted on Synapse and mounted in `/modelState`.

To summarize, I recommend to provide a scratch space of 500 GB to inference method because less than 10 GB is a bit short these days and for this application. Let's keep the maximum docker size to 10 GB.

@brucehoff Participants saves their predictions to `/output/predictions.tsv`. What is the partition behind `/output/` and how much data can they write to it or `/output/predictions.tsv`?

**Comment by brucehoff** [comment]
@tschaffter is it the *container* or the *image* that's limited to 10GB (or both)?

> What is the partition behind /output/

Currently it's `/root/tempDirX`, X=0,1.  You have proposed that we switch to a dedicated partition `/data/tempDirX`, https://github.com/Sage-Bionetworks/DigitalMammographyChallenge/issues/604

**Comment by tschaffter** [comment]
> is it the container or the image that's limited to 10GB (or both)?

By default, both.

> Currently it's /root/tempDirX, X=0,1. You have proposed that we switch to a dedicated partition /data/tempDirX, #604

Oh! Yes, we have to limit the ability of the user to write. Currently they could crash a machine by writing a large amount of data to `/root/tempDirX`. As discussed, the agents are going to use a temp directory on a dedicated partition.

@brucehoff What happens if the temp directory of an agent saturate? Does the agent crash? When running the inference submission, do you mount only an empty file `/output/predictions.tsv` or do you mount `/root/tempDir` (not safe if the agent writes its own files there) or a subdirectory to `/output/`?

**Comment by brucehoff** [comment]
> @brucehoff What happens if the temp directory of an agent saturates? Does the agent crash? 
Yes, I think it would:  The use of the temp dir is to zip up the logs or model state for upload.  If the zip operation fails due to lack of disk space then an exception would be thrown.  The agent is designed to send an email and then exit.  (Don't know if this is exactly "crashing", but in any case the agent will halt under such a condition.)  I do not believe this has ever actually happened.

> do you mount only an empty file /output/predictions.tsv or do you mount /root/tempDir

We create a subdirectory of `/root/tempDirX/`, e.g., `/root/tempDir0/<UUID>/` and mount it as `/output/`.  I had not thought to mount an empty, writable file (e.g. mount a *file* `/root/tempDir0/<UUID>/predictions.tsv` as `/output/predictions.tsv`).  That would certainly be cleaner but I've never tried it and I don't know if you can mount a single writable file (rather than a volume) using the `-v` option.  It might be worth trying.

**Comment by tschaffter** [comment]
@brucehoff Yes, docker can mount a single file as "volume" (e.g. configuration files). Because an agent will shutdown if a participant write too much data to `/output/`, I propose to mount a dedicated partition `/data/inferenceX/` on the host to `/output/` in the container. Here X has the same value than the corresponding agent index. The new partitions would be relatively small (50 MB or so, more than enough to host predictions.txt). Do you agree?

**Comment by brucehoff** [comment]
Yes, that sounds like a great idea.  I am learning that anytime we provide write access we must set a maximum size or else the system will not be stable.  This is a perfect example.

**Comment by tschaffter** [comment]
@brucehoff Do you already mount the scratch space that is now present on the inference machines (will be created on the SoftLayer inference machines when they will be updated)? If yes, this ticket can be closed.

**Comment by brucehoff** [comment]
@tschaffter Not yet, no.

**Comment by brucehoff** [comment]
https://github.com/Sage-Bionetworks/ChallengeDockerAgent/commit/52a5635f57a87598b967fe91c8023dd0294a4fcc

**Comment by brucehoff** [comment]
Remaining task is to update the amazon inference machines.

**Comment by brucehoff** [comment]
The 10 amazon machines have now been updated (also the express lane).
